### ISSUE TITLE
"-all" classifier dropped from published jar

### ISSUE THREAD

**Issue by HexagonSun** [GradleUp/shadow#860]
When publishing a Maven artifact using Shadow 8.1.1, the "-all" classifier is dropped from the jar file. 

This happens for both local publishing and to remote repositories, too. 
  
### Shadow Version
8.1.1

### Gradle Version
8.1

### Expected Behavior
Building using Shadow 8.1.0
  ```
    rm -rf build && rm -rf ${local-maven-repo} && gw publishToMavenLocal
  ```
produces the following artifacts:
  ```
    /Users/hexagonsun/.m2/repository/net/hexagon/sun/foo$find .
    .
    ./dev-local
    ./dev-local/foo-dev-local.pom
    ./dev-local/foo-dev-local-all.jar
    ./maven-metadata-local.xml
  ```
### Actual Behavior
Building with Shadow 8.1.1
  ```
    rm -rf build && rm -rf ${local-maven-repo} && gw publishToMavenLocal
  ```
produces the following artifacts:
  ```
    /Users/hexagonsun/.m2/repository/net/hexagon/sun/foo$find .
    .
    ./dev-local
    ./dev-local/foo-dev-local.pom
    ./dev-local/foo-dev-local.jar
    ./maven-metadata-local.xml
  ```
Note that the jar is lacking the `-all` qualifier.

In both cases, i.e. with Shadow 8.1.0 as well as the new 8.1.1, the shadow-jar of the application's `build` directory correctly contains the `-all` classifier:
  ```
    /Users/hexagonsun/dev/foo/build$find . -iname "*.jar"
    ./libs/foo-dev-local-all.jar
  ```

### Gradle Build Script(s)
```
import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar

plugins {
    val kotlinVersion = "1.8.20"

    application
    `maven-publish`
    kotlin("jvm") version kotlinVersion
    id("com.github.johnrengelman.shadow") version "8.1.1"
}

group = "net.hexagon.sun.foo"
version = "dev-local"

repositories {
    val mavenCentralProxyUrl: String by properties
    maven { url = uri(mavenCentralProxyUrl) }
}

dependencies {
    implementation(kotlin("stdlib"))
}

application {
    mainClass.set("net.hexagon.sun.foo.ApplicationKt")
}

tasks {
    withType<ShadowJar> {
        archiveClassifier.set("all")
    }
}

publishing {
    publications {
        create<MavenPublication>("shadow") {
            project.shadow.component(this)
        }
    }
}

```

Note that the line `archiveClassifier.set("all")` is not relevant: the same problem occurs when it is missing.

### Content of Shadow JAR (`jar tf <jar file>` - post link to GIST if too long)

I don't think this is relevant, since the content is fine. Let me know if I should supply it.
The behavior can be reproduced using a default `Application.kt` file.
```
package net.hexagon.sun.foo

fun main() {
}
```

**Comment by HexagonSun** [comment]
I've tracked this down to changes in 960617f..a6399e2 (the intermediate commits produce errors). 

At a glance, I don't see any obvious things that caused the classifier do be omitted, then again I'm not familiar at all with this codebase. @Goooler the changes were made by you, do you know what could have caused this issue?

**Comment by Goooler** [comment]
It works well in https://github.com/pinterest/ktlint/blob/master/ktlint-cli/build.gradle.kts

**Comment by HexagonSun** [comment]
As far as I can tell `tasks.shadowJar.orNull?.outputs?.files?.singleFile` uses the file from the local `build` directory, which works indeed fine (as I initially stated, too). 

The problem only arises when publishing to local or remote repositories. I'm _guessing_ this means that `project.shadow.component(this)` is somehow involved.

**Comment by tha00** [comment]
In 960617f, the artifact addition was changed `ShadowExtension` from
```groovy
publication.artifact(project.tasks.named("shadowJar"))
```
to
```groovy
publication.artifact(archiveFile.get())
```
where `archiveFile` provides `project.tasks.withType(ShadowJar).getByName("shadowJar").archiveFile`.

The docs of [MavenPublication#artifact](https://docs.gradle.org/current/javadoc/org/gradle/api/publish/maven/MavenPublication.html#artifact-java.lang.Object-) give different behaviour for different input types:

> * A [PublishArtifact](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/PublishArtifact.html) instance. Extension and classifier values are taken from the wrapped instance.
> * Anything that can be resolved to a File via the [Project.file(Object)](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html#file-java.lang.Object-) method. Extension and classifier values are interpolated from the file name.

Could it be that the classifier is not interpolated correctly from the file name?

**Comment by Goooler** [comment]
I've published the fix from #904 in my fork, you can try https://plugins.gradle.org/plugin/io.github.goooler.shadow
